sql/2024/04/inaccurate-sizes-attribute-impact.sql

```diff
@@ -0,0 +1,90 @@
+# HTTP Archive query to measure impact of inaccurate sizes attributes per <img> for WordPress sites.
+#
+# WPP Research, Copyright 2022 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+CREATE TEMPORARY FUNCTION
+  getImgSizesAccuracy(custom_metrics STRING)
+  RETURNS ARRAY<STRUCT<sizesAbsoluteError INT64,
+  sizesRelativeError FLOAT64,
+  idealSizesSelectedResourceEstimatedPixels INT64,
+  actualSizesEstimatedWastedLoadedPixels INT64,
+  idealSizesSelectedResourceEstimatedBytes FLOAT64,
+  actualSizesEstimatedWastedLoadedBytes FLOAT64>>
+  LANGUAGE js AS '''
+try {
+  const $ = JSON.parse(custom_metrics);
+  let responsiveImages = JSON.parse($.responsive_images);
+  responsiveImages = responsiveImages['responsive-images'];
+
+  return responsiveImages.map((imgData) => {
+    return {
+      imgData.sizesAbsoluteError,
+      imgData.sizesRelativeError,
+      imgData.idealSizesSelectedResourceEstimatedPixels
+      imgData.actualSizesEstimatedWastedLoadedPixels,
+      imgData.idealSizesSelectedResourceEstimatedBytes
+      imgData.actualSizesEstimatedWastedLoadedBytes,
```

Comment:
Running some test queries using this script for a specific origin, I noticed it includes several results where each property is `NULL`. That makes sense because some images don't even have `srcset` or `sizes` attributes. So we need to make sure to ignore them here.

FWIW, the original query probably has this problem too, we just never noticed that.

I think the easiest solution would be to also return the `hasSrcset` and `hasSizes` properties here, and then in the `wordpressSizesData` query below you can add `image.hasSrcset AND image.hasSizes` to the `WHERE` clause to filter correctly.

Reply:
Good idea. Added in [87964b0](https://github.com/GoogleChromeLabs/wpp-research/pull/108/commits/87964b0123bdf8dbfd006288937a13bfae8aae3a).